source/shared/smart-contracts/tutorials/piggy-bank/testing.rst

```diff
@@ -226,74 +227,75 @@ Run the test to check that it compiles and succeeds.
+When calling ``piggy_insert`` you get back a result with a return value as
+opposed to the initial state that you get from ``piggy_init``:
 
 .. code-block:: rust
 
-   let actions_result: ReceiveResult<ActionsTree> = piggy_insert(&ctx, amount, &mut state);
+   let result = piggy_insert(&ctx, &host, amount);
 
-For testing you can represent the actions as a simple tree structure
-|ActionsTree|_, making it easy to inspect.
+The test then checks whether the result was ok:
 
-.. note::
+.. code-block:: rust
 
-   The |receive| macro uses another representation of the actions, when building
-   the smart contract module. This representation depends on functions supplied
-   by the host environment and is therefore not suitable for unit tests.
+   assert!(result.is_ok(), "Inserting CCD results in error");
 
-Now you need to check if the function succeeded and verify the resulting state and actions.
-In this case the state should remain intact and the function produce only the action for accepting the CCD.
+One test that is tempting to add is to check that the piggy bank remains intact
+after inserting CCD into it:
 
 .. code-block:: rust
 
-   let actions = actions_result.expect("Inserting CCD results in error.");
+   assert_eq!(
+       *host.state(),
+       PiggyBankState::Intact,
+       "Piggy bank state should still be intact."
+   );
 
-   assert_eq!(actions, ActionsTree::accept(), "No action should be produced.");
-   assert_eq!(state, PiggyBankState::Intact, "Piggy bank state should still be intact.");
+However, there is no way for the immutable receive method ``piggy_insert`` to
+mutate the state.
+Trying to do so would result in an error from the Rust compiler.
+By using immutable receive functions, it is possible to rule out certain error
+cases at compile time, which means that we do not need tests for these
+scenarios.
+Along with performance, those are the two primary reasons for not making your
+receive methods |mutable|_ unless strictly necessary.
 
 The second test becomes:
 
 .. code-block:: rust
 
    #[test]
    fn test_insert_intact() {
-       let ctx = ReceiveContextTest::empty();
+       let ctx = TestReceiveContext::empty();
+       let host = TestHost::new(PiggyBankState::Intact);
        let amount = Amount::from_micro_ccd(100);
-       let mut state = PiggyBankState::Intact;
-
-       let actions_result: ReceiveResult<ActionsTree> = piggy_insert(&ctx, amount, &mut state);
 
-       let actions = actions_result.expect("Inserting CCD results in error.");
+       let result = piggy_insert(&ctx, &host, amount);
 
-       assert_eq!(actions, ActionsTree::accept(), "No action should be produced.");
-       assert_eq!(state, PiggyBankState::Intact, "Piggy bank state should still be intact.");
+       assert!(result.is_ok(), "Inserting CCD results in error");
    }
 
-Again you should verify everything compiles and the tests succeeds using ``cargo
+Again, you should verify everything compiles and the tests succeeds using ``cargo
```

Comment:
```suggestion
Again, you should verify everything compiles and the test succeeds using ``cargo
```

Reply:
`cargo test` will run all the tests, so I believe `tests` is correct at this point 😊